src/python-bindings/htcondor2/job_event_log.cpp

```diff
@@ -1 +1,219 @@
+        return NULL;
+    }
+    handle->f = [](void *& v) { dprintf( D_PERF_TRACE, "[WaitForUserLog]\n" ); delete (WaitForUserLog *)v; v = NULL; };
+    handle->t = (void *)wful;
+
+    Py_RETURN_NONE;
+}
+
+
+#ifdef WIN32
+    #include <windows.h>
+    #define MODULE_LOCK_MUTEX_TYPE CRITICAL_SECTION
+    #define MODULE_LOCK_MUTEX_LOCK EnterCriticalSection
+    #define MODULE_LOCK_MUTEX_UNLOCK LeaveCriticalSection
+    #define MODULE_LOCK_MUTEX_INITIALIZE(mtx) InitializeCriticalSection(mtx)
+#else
+    #include <pthread.h>
+    #define MODULE_LOCK_MUTEX_TYPE pthread_mutex_t
+    #define MODULE_LOCK_MUTEX_INITIALIZER PTHREAD_MUTEX_INITIALIZER
+    #define MODULE_LOCK_MUTEX_LOCK pthread_mutex_lock
+    #define MODULE_LOCK_MUTEX_UNLOCK pthread_mutex_unlock
+#endif
+
+
+class JobEventLogGlobalLockInitializer {
+    public:
+        JobEventLogGlobalLockInitializer() {
+#if defined(WIN32)
+            MODULE_LOCK_MUTEX_INITIALIZE( & mutex );
+#else
+            mutex = MODULE_LOCK_MUTEX_INITIALIZER;
+#endif
+        }
+
+        MODULE_LOCK_MUTEX_TYPE mutex;
+} jobEventLogGlobalLock;
+
+
+static PyObject *
+_job_event_log_next( PyObject *, PyObject * args ) {
+    PyObject * self = NULL;
+    PyObject_Handle * handle = NULL;
+    long deadline = 0;
+
+    if(! PyArg_ParseTuple( args, "OOl", & self, (PyObject **)& handle, & deadline )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    auto wful = (WaitForUserLog *)handle->t;
+
+
+    ULogEvent * event = NULL;
+    // Must not be declared inside the invisible scope in which we allow
+    // other Python threads to run.
+    ULogEventOutcome outcome;
+
+    Py_BEGIN_ALLOW_THREADS
+    MODULE_LOCK_MUTEX_LOCK( & jobEventLogGlobalLock.mutex );
+
+        if( deadline ) {
+            time_t now = time(NULL);
+
+            if( deadline <= now ) {
+                // If the deadline is now, or has passed, poll for an event.
+                outcome = wful->readEvent( event, 0, false );
+            } else {
+                // If the deadline has yet to pass, block until it does.
+                outcome = wful->readEvent( event, (deadline - now) * 1000, true );
+            }
+        } else {
+            // If there isn't a deadline, block forever.
+            outcome = wful->readEvent( event, -1, true );
+        }
+
+    MODULE_LOCK_MUTEX_UNLOCK( & jobEventLogGlobalLock.mutex );
+    Py_END_ALLOW_THREADS
+
+    switch( outcome ) {
+        case ULOG_OK: {
+            // Obtain the serialized form of the event.
+            std::string event_text;
+
+            int fo = USERLOG_FORMAT_DEFAULT;
+            auto format = param( "DEFAULT_USERLOG_FORMAT_OPTIONS" );
+            if( format != NULL ) {
+                fo = ULogEvent::parse_opts( format, USERLOG_FORMAT_DEFAULT );
+                free( format );
+                format = NULL;
+            }
+
+            if(! event->formatEvent( event_text, fo )) {
```

Comment:
It'd sure be nice if we could just capture the text that was read from the user log instead of reconstituting it. I'm not proposing that we do that in this PR.